api4/ldap.go

```diff
@@ -38,3 +49,167 @@ func testLdap(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	ReturnStatusOK(w)
 }
+
+func getLdapGroups(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	if c.App.License() == nil || !*c.App.License().Features.LDAPGroups {
+		c.Err = model.NewAppError("Api4.getLdapGroups", "api.ldap.license.error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	groups, total, err := c.App.GetAllLdapGroupsPage(c.Params.Page, c.Params.PerPage)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if len(groups) == 0 {
+		w.Write([]byte("[]"))
+		return
+	}
+
+	mugs := []*model.MixedUnlinkedGroup{}
+	for _, group := range groups {
+		mug := &model.MixedUnlinkedGroup{
+			DisplayName: group.DisplayName,
+			RemoteId:    group.RemoteId,
+		}
+		if len(group.Id) == 26 {
+			mug.Id = &group.Id
+			mug.HasSyncables = &group.HasSyncables
+		}
+		mugs = append(mugs, mug)
+	}
+
+	b, marshalErr := json.Marshal(struct {
+		Count  int                         `json:"count"`
+		Groups []*model.MixedUnlinkedGroup `json:"groups"`
+	}{Count: total, Groups: mugs})
+	if marshalErr != nil {
+		c.Err = model.NewAppError("Api4.getLdapGroups", "api.ldap.marshal_error", nil, marshalErr.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Write(b)
+}
+
+func linkLdapGroup(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireRemoteId()
+	if c.Err != nil {
+		return
+	}
+
+	if !c.App.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	if c.App.License() == nil || !*c.App.License().Features.LDAPGroups {
+		c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.license.error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	ldapGroup, err := c.App.GetLdapGroup(c.Params.RemoteId)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	if ldapGroup == nil {
+		c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.not_found", nil, "", http.StatusNotFound)
+		return
+	}
+
+	group, err := c.App.GetGroupByRemoteID(ldapGroup.RemoteId, model.GroupTypeLdap)
+	if err != nil && err.DetailedError != sql.ErrNoRows.Error() {
+		c.Err = err
+		return
+	}
+
+	var status int
+	var newOrUpdatedGroup *model.Group
+
+	// Group is already linked.
+	if group != nil {
+		if group.DeleteAt == 0 {
+			c.Err = model.NewAppError("Api4.linkLdapGroup", "api.ldap.already_linked", nil, "", http.StatusNotModified)
+			return
```

Comment:
Is this an error case? If so, I don't think `StatusNotModified` is appropriate here. Or is the idea that it's basically an `It's OK, you already did this` response.

Reply:
It's the latter. "You've already linked this so there's no further changes that can be made" kind of thing.

Reply:
OK makes sense. Does returning a 304 have any other side effects? I'm more used to seeing it in relation to "use the cached version of this resource", which feels slightly different to what is meant here.

Reply:
I actually think you're right [according to MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304)
> It is an implicit redirection to a cached resource.
I'll change it.

Reply:
I ended up switching removing the "already linked" and "already unlinked" cases from error responses to success responses.